File: ios/link2ur/link2ur/Views/Profile/TaskPreferencesView.swift
import SwiftUI
import Combine

struct TaskPreferencesView: View {
    @EnvironmentObject var appState: AppState
    @Environment(\.dismiss) var dismiss
    @StateObject private var viewModel = TaskPreferencesViewModel()
    
    // 任务类型选项
    private var taskTypes: [(name: String, value: String)] {
        [
            (LocalizationKey.taskCategoryHousekeeping.localized, "Housekeeping"),
            (LocalizationKey.taskCategoryCampusLife.localized, "Campus Life"),
            (LocalizationKey.taskCategorySecondhandRental.localized, "Second-hand & Rental"),
            (LocalizationKey.taskCategoryErrandRunning.localized, "Errand Running"),
            (LocalizationKey.taskCategorySkillService.localized, "Skill Service"),
            (LocalizationKey.taskCategorySocialHelp.localized, "Social Help"),
            (LocalizationKey.taskCategoryTransportation.localized, "Transportation"),
            (LocalizationKey.taskCategoryPetCare.localized, "Pet Care"),
            (LocalizationKey.taskCategoryLifeConvenience.localized, "Life Convenience"),
            (LocalizationKey.taskCategoryOther.localized, "Other")
        ]
    }
    
    // 地点选项
    private let locations = [
        "Online", "London", "Edinburgh", "Manchester", "Birmingham", "Glasgow",
        "Bristol", "Sheffield", "Leeds", "Nottingham", "Newcastle", "Southampton",
        "Liverpool", "Cardiff", "Coventry", "Exeter", "Leicester", "York",
        "Aberdeen", "Bath", "Dundee", "Reading", "St Andrews", "Belfast",
        "Brighton", "Durham", "Norwich", "Swansea", "Loughborough", "Lancaster",
        "Warwick", "Cambridge", "Oxford", "Other"
    ]
    
    // 任务等级选项
    private let taskLevels = ["Normal", "VIP", "Super"]
    
    var body: some View {
        NavigationStack {
            ZStack {
                AppColors.background
                    .ignoresSafeArea()
                
                if viewModel.isLoading && viewModel.preferences == nil {
                    ProgressView()
                } else {
                    ScrollView {
                        VStack(spacing: AppSpacing.xl) {
                            // 偏好的任务类型
                            PreferenceSection(
                                title: LocalizationKey.taskPreferencesPreferredTypes.localized,
                                description: LocalizationKey.taskPreferencesPreferredTypesDescription.localized
                            ) {
                                LazyVGrid(columns: [
                                    GridItem(.flexible(), spacing: AppSpacing.sm),
                                    GridItem(.flexible(), spacing: AppSpacing.sm)
                                ], spacing: AppSpacing.sm) {
                                    ForEach(taskTypes, id: \.value) { type in
                                        PreferenceToggle(
                                            title: type.name,
                                            isSelected: viewModel.preferences?.taskTypes.contains(type.value) ?? false
                                        ) {
                                            viewModel.toggleTaskType(type.value)
                                        }
                                    }
                                }
                            }
                            
                            // 偏好的地点
                            PreferenceSection(
                                title: LocalizationKey.taskPreferencesPreferredLocations.localized,
                                description: LocalizationKey.taskPreferencesPreferredLocationsDescription.localized
                            ) {
                                LazyVGrid(columns: [
                                    GridItem(.flexible(), spacing: AppSpacing.sm),
                                    GridItem(.flexible(), spacing: AppSpacing.sm)
                                ], spacing: AppSpacing.sm) {
                                    ForEach(locations, id: \.self) { location in
                                        PreferenceToggle(
                                            title: location,
                                            isSelected: viewModel.preferences?.locations.contains(location) ?? false
                                        ) {
                                            viewModel.toggleLocation(location)
                                        }
                                    }
                                }
                            }
                            
                            // 偏好的任务等级
                            PreferenceSection(
                                title: LocalizationKey.taskPreferencesPreferredLevels.localized,
                                description: LocalizationKey.taskPreferencesPreferredLevelsDescription.localized
                            ) {
                                HStack(spacing: AppSpacing.md) {
                                    ForEach(taskLevels, id: \.self) { level in
                                        PreferenceToggle(
                                            title: level,
                                            isSelected: viewModel.preferences?.taskLevels.contains(level) ?? false
                                        ) {
                                            viewModel.toggleTaskLevel(level)
                                        }
                                    }
                                }
                            }
                            
                            // 最少截止时间
                            PreferenceSection(
                                title: LocalizationKey.taskPreferencesMinDeadline.localized,
                                description: LocalizationKey.taskPreferencesMinDeadlineDescription.localized
                            ) {
                                HStack(spacing: AppSpacing.md) {
                                    Stepper(
                                        value: Binding(
                                            get: { viewModel.preferences?.minDeadlineDays ?? 1 },
                                            set: { viewModel.updateMinDeadlineDays($0) }
                                        ),
                                        in: 1...30
                                    ) {
                                        Text("\(viewModel.preferences?.minDeadlineDays ?? 1) \(LocalizationKey.taskPreferencesDays.localized)")
                                            .font(AppTypography.body)
                                            .foregroundColor(AppColors.textPrimary)
                                    }
                                    
                                    Text(LocalizationKey.taskPreferencesDaysRange.localized)
                                        .font(AppTypography.caption)
                                        .foregroundColor(AppColors.textSecondary)
                                }
                            }
                            
                            // 保存按钮
                            Button {
                                viewModel.savePreferences {
                                    dismiss()
                                }
                            } label: {
                                Text(LocalizationKey.taskPreferencesSave.localized)
                                    .font(AppTypography.bodyBold)
                                    .foregroundColor(.white)
                                    .frame(maxWidth: .infinity)
                                    .frame(height: 56)
                                    .background(
                                        LinearGradient(
                                            gradient: Gradient(colors: AppColors.gradientPrimary),
                                            startPoint: .leading,
                                            endPoint: .trailing
                                        )
                                    )
                                    .cornerRadius(AppCornerRadius.large)
                                    .shadow(color: AppColors.primary.opacity(0.3), radius: 12, x: 0, y: 6)
                            }
                            .buttonStyle(PrimaryButtonStyle(cornerRadius: AppCornerRadius.large, useGradient: true))
                            .padding(.horizontal, AppSpacing.md)
                            .padding(.bottom, AppSpacing.xl)
                            .disabled(viewModel.isSaving)
                        }
                        .padding(.horizontal, AppSpacing.md)
                        .padding(.top, AppSpacing.md)
                    }
                }
            }
            .navigationTitle(LocalizationKey.taskPreferencesTitle.localized)
            .navigationBarTitleDisplayMode(.inline)
            .enableSwipeBack()
            .toolbar {
                ToolbarItem(placement: .navigationBarTrailing) {
                    Button(LocalizationKey.commonDone.localized) {
                        dismiss()
                    }
                    .foregroundColor(AppColors.primary)
                }
            }
            .onAppear {
                viewModel.loadPreferences()
            }
        }
    }
}

struct PreferenceSection<Content: View>: View {
    let title: String
    let description: String
    @ViewBuilder let content: Content
    
    var body: some View {
        VStack(alignment: .leading, spacing: AppSpacing.md) {
            Text(title)
                .font(AppTypography.title3)
                .fontWeight(.bold)
                .foregroundColor(AppColors.textPrimary)
            
            Text(description)
                .font(AppTypography.caption)
                .foregroundColor(AppColors.textSecondary)
            
            content
        }
        .padding(AppSpacing.md)
        .background(AppColors.cardBackground)
        .cornerRadius(AppCornerRadius.large)
        .shadow(color: AppShadow.small.color, radius: AppShadow.small.radius, x: AppShadow.small.x, y: AppShadow.small.y)
    }
}

struct PreferenceToggle: View {
    let title: String
    let isSelected: Bool
    let action: () -> Void
    
    var body: some View {
        Button(action: action) {
            Text(title)
                .font(AppTypography.body)
                .foregroundColor(isSelected ? AppColors.primary : AppColors.textPrimary)
                .frame(maxWidth: .infinity)
                .padding(.vertical, AppSpacing.sm)
                .background(isSelected ? AppColors.primary.opacity(0.1) : AppColors.cardBackground)
                .cornerRadius(AppCornerRadius.medium)
                .overlay(
                    RoundedRectangle(cornerRadius: AppCornerRadius.medium)
                        .stroke(isSelected ? AppColors.primary : AppColors.separator, lineWidth: isSelected ? 2 : 1)
                )
        }
        .buttonStyle(ScaleButtonStyle())
    }
}

@MainActor
class TaskPreferencesViewModel: ObservableObject {
    @Published var preferences: UserPreferences?
    @Published var isLoading = false
    @Published var isSaving = false
    @Published var errorMessage: String?
    
    private let apiService = APIService.shared
    private var cancellables = Set<AnyCancellable>()
    
    func loadPreferences() {
        isLoading = true
        errorMessage = nil
        
        apiService.getUserPreferences()
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    self?.isLoading = false
                    if case .failure(let error) = completion {
                        self?.errorMessage = error.userFriendlyMessage
                        // 如果加载失败，使用默认值
                        self?.preferences = UserPreferences()
                    }
                },
                receiveValue: { [weak self] prefs in
                    self?.isLoading = false
                    self?.preferences = prefs
                }
            )
            .store(in: &cancellables)
    }
    
    func toggleTaskType(_ type: String) {
        guard let prefs = preferences else { return }
        if prefs.taskTypes.contains(type) {
            preferences = UserPreferences(
                taskTypes: prefs.taskTypes.filter { $0 != type },
                locations: prefs.locations,
                taskLevels: prefs.taskLevels,
                keywords: prefs.keywords,
                minDeadlineDays: prefs.minDeadlineDays
            )
        } else {
            preferences = UserPreferences(
                taskTypes: prefs.taskTypes + [type],
                locations: prefs.locations,
                taskLevels: prefs.taskLevels,
                keywords: prefs.keywords,
                minDeadlineDays: prefs.minDeadlineDays
            )
        }
    }
    
    func toggleLocation(_ location: String) {
        guard let prefs = preferences else { return }
        if prefs.locations.contains(location) {
            preferences = UserPreferences(
                taskTypes: prefs.taskTypes,
                locations: prefs.locations.filter { $0 != location },
                taskLevels: prefs.taskLevels,
                keywords: prefs.keywords,
                minDeadlineDays: prefs.minDeadlineDays
            )
        } else {
            preferences = UserPreferences(
                taskTypes: prefs.taskTypes,
                locations: prefs.locations + [location],
                taskLevels: prefs.taskLevels,
                keywords: prefs.keywords,
                minDeadlineDays: prefs.minDeadlineDays
            )
        }
    }
    
    func toggleTaskLevel(_ level: String) {
        guard let prefs = preferences else { return }
        if prefs.taskLevels.contains(level) {
            preferences = UserPreferences(
                taskTypes: prefs.taskTypes,
                locations: prefs.locations,
                taskLevels: prefs.taskLevels.filter { $0 != level },
                keywords: prefs.keywords,
                minDeadlineDays: prefs.minDeadlineDays
            )
        } else {
            preferences = UserPreferences(
                taskTypes: prefs.taskTypes,
                locations: prefs.locations,
                taskLevels: prefs.taskLevels + [level],
                keywords: prefs.keywords,
                minDeadlineDays: prefs.minDeadlineDays
            )
        }
    }
    
    func updateMinDeadlineDays(_ days: Int) {
        guard let prefs = preferences else { return }
        preferences = UserPreferences(
            taskTypes: prefs.taskTypes,
            locations: prefs.locations,
            taskLevels: prefs.taskLevels,
            keywords: prefs.keywords,
            minDeadlineDays: days
        )
    }
    
    func savePreferences(completion: @escaping () -> Void) {
        guard let prefs = preferences else { return }
        
        isSaving = true
        errorMessage = nil
        
        apiService.updateUserPreferences(preferences: prefs)
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    self?.isSaving = false
                    if case .failure(let error) = completion {
                        self?.errorMessage = error.userFriendlyMessage
                    }
                },
                receiveValue: { [weak self] _ in
                    self?.isSaving = false
                    HapticFeedback.success()
                    completion()
                }
            )
            .store(in: &cancellables)
    }
}

File: link2ur/lib/features/profile/views/task_preferences_view.dart
import 'package:flutter/material.dart';
import 'package:flutter_bloc/flutter_bloc.dart';

import '../../../core/design/app_colors.dart';
import '../../../core/design/app_spacing.dart';
import '../../../core/design/app_radius.dart';
import '../../../core/utils/l10n_extension.dart';
import '../../../data/repositories/user_repository.dart';

/// 任务偏好设置页
/// 参考iOS TaskPreferencesView.swift
class TaskPreferencesView extends StatefulWidget {
  const TaskPreferencesView({super.key});

  @override
  State<TaskPreferencesView> createState() => _TaskPreferencesViewState();
}

class _TaskPreferencesViewState extends State<TaskPreferencesView> {
  bool _isLoading = true;
  bool _isSaving = false;

  final Set<String> _selectedTaskTypes = {};
  final Set<String> _selectedLocations = {};
  final Set<String> _selectedLevels = {};
  int _minDeadlineDays = 1;

  static const _taskTypes = [
    ('Housekeeping', 'housekeeping'),
    ('Campus Life', 'campusLife'),
    ('Second-hand & Rental', 'secondhandRental'),
    ('Errand Running', 'errandRunning'),
    ('Skill Service', 'skillService'),
    ('Social Help', 'socialHelp'),
    ('Transportation', 'transportation'),
    ('Pet Care', 'petCare'),
    ('Life Convenience', 'lifeConvenience'),
    ('Other', 'other'),
  ];

  static const _locations = [
    'Online', 'London', 'Edinburgh', 'Manchester', 'Birmingham',
    'Glasgow', 'Bristol', 'Sheffield', 'Leeds', 'Nottingham',
    'Newcastle', 'Southampton', 'Liverpool', 'Cardiff', 'Coventry',
    'Exeter', 'Leicester', 'York', 'Aberdeen', 'Bath', 'Dundee',
    'Reading', 'St Andrews', 'Belfast', 'Brighton', 'Durham',
    'Norwich', 'Swansea', 'Loughborough', 'Lancaster', 'Warwick',
    'Cambridge', 'Oxford', 'Other',
  ];

  static const _taskLevels = ['Normal', 'VIP', 'Super'];

  @override
  void initState() {
    super.initState();
    _loadPreferences();
  }

  Future<void> _loadPreferences() async {
    try {
      final repo = context.read<UserRepository>();
      final prefs = await repo.getUserPreferences();
      if (mounted) {
        setState(() {
          _selectedTaskTypes.addAll(prefs['task_types'] as List<String>? ?? []);
          _selectedLocations.addAll(prefs['locations'] as List<String>? ?? []);
          _selectedLevels.addAll(prefs['task_levels'] as List<String>? ?? []);
          _minDeadlineDays = prefs['min_deadline_days'] as int? ?? 1;
          _isLoading = false;
        });
      }
    } catch (_) {
      if (mounted) setState(() => _isLoading = false);
    }
  }

  Future<void> _savePreferences() async {
    setState(() => _isSaving = true);

    try {
      final repo = context.read<UserRepository>();
      await repo.updateUserPreferences({
        'task_types': _selectedTaskTypes.toList(),
        'locations': _selectedLocations.toList(),
        'task_levels': _selectedLevels.toList(),
        'min_deadline_days': _minDeadlineDays,
      });
      if (mounted) Navigator.of(context).pop();
    } catch (e) {
      if (mounted) {
        setState(() => _isSaving = false);
        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(content: Text(e.toString())),
        );
      }
    }
  }

  void _toggleItem(Set<String> set, String item) {
    setState(() {
      if (set.contains(item)) {
        set.remove(item);
      } else {
        set.add(item);
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    final l10n = context.l10n;

    return Scaffold(
      appBar: AppBar(
        title: Text(l10n.taskPreferencesTitle),
        actions: [
          TextButton(
            onPressed: () => Navigator.of(context).pop(),
            child: Text(l10n.commonDone),
          ),
        ],
      ),
      body: _isLoading
          ? const Center(child: CircularProgressIndicator())
          : SingleChildScrollView(
              padding: const EdgeInsets.all(AppSpacing.md),
              child: Column(
                children: [
                  // 偏好任务类型
                  _PreferenceSection(
                    title: l10n.taskPreferencesTypes,
                    description: l10n.taskPreferencesTypesDesc,
                    child: Wrap(
                      spacing: AppSpacing.sm,
                      runSpacing: AppSpacing.sm,
                      children: _taskTypes.map((type) {
                        final isSelected =
                            _selectedTaskTypes.contains(type.$1);
                        return _ToggleChip(
                          label: type.$1,
                          isSelected: isSelected,
                          onTap: () =>
                              _toggleItem(_selectedTaskTypes, type.$1),
                        );
                      }).toList(),
                    ),
                  ),
                  const SizedBox(height: AppSpacing.lg),

                  // 偏好地点
                  _PreferenceSection(
                    title: l10n.taskPreferencesLocations,
                    description: l10n.taskPreferencesLocationsDesc,
                    child: Wrap(
                      spacing: AppSpacing.sm,
                      runSpacing: AppSpacing.sm,
                      children: _locations.map((loc) {
                        final isSelected =
                            _selectedLocations.contains(loc);
                        return _ToggleChip(
                          label: loc,
                          isSelected: isSelected,
                          onTap: () =>
                              _toggleItem(_selectedLocations, loc),
                        );
                      }).toList(),
                    ),
                  ),
                  const SizedBox(height: AppSpacing.lg),

                  // 偏好等级
                  _PreferenceSection(
                    title: l10n.taskPreferencesLevels,
                    description: l10n.taskPreferencesLevelsDesc,
                    child: Row(
                      children: _taskLevels.map((level) {
                        final isSelected =
                            _selectedLevels.contains(level);
                        return Expanded(
                          child: Padding(
                            padding: const EdgeInsets.symmetric(
                                horizontal: 4),
                            child: _ToggleChip(
                              label: level,
                              isSelected: isSelected,
                              onTap: () =>
                                  _toggleItem(_selectedLevels, level),
                            ),
                          ),
                        );
                      }).toList(),
                    ),
                  ),
                  const SizedBox(height: AppSpacing.lg),

                  // 最短截止时间
                  _PreferenceSection(
                    title: l10n.taskPreferencesMinDeadline,
                    description: l10n.taskPreferencesMinDeadlineDesc,
                    child: Row(
                      children: [
                        Text(
                          '$_minDeadlineDays ${l10n.taskPreferencesDays}',
                          style: const TextStyle(
                              fontSize: 16,
                              fontWeight: FontWeight.w600),
                        ),
                        const Spacer(),
                        IconButton(
                          onPressed: _minDeadlineDays > 1
                              ? () => setState(
                                  () => _minDeadlineDays--)
                              : null,
                          icon: const Icon(Icons.remove_circle_outline),
                        ),
                        IconButton(
                          onPressed: _minDeadlineDays < 30
                              ? () => setState(
                                  () => _minDeadlineDays++)
                              : null,
                          icon: const Icon(Icons.add_circle_outline),
                        ),
                      ],
                    ),
                  ),
                  const SizedBox(height: AppSpacing.xl),

                  // 保存按钮
                  SizedBox(
                    width: double.infinity,
                    height: 56,
                    child: ElevatedButton(
                      onPressed: _isSaving ? null : _savePreferences,
                      style: ElevatedButton.styleFrom(
                        backgroundColor: AppColors.primary,
                        foregroundColor: Colors.white,
                        shape: RoundedRectangleBorder(
                          borderRadius:
                              BorderRadius.circular(AppRadius.large),
                        ),
                      ),
                      child: _isSaving
                          ? const SizedBox(
                              width: 24,
                              height: 24,
                              child: CircularProgressIndicator(
                                  color: Colors.white, strokeWidth: 2))
                          : Text(l10n.taskPreferencesSave,
                              style: const TextStyle(
                                  fontSize: 16,
                                  fontWeight: FontWeight.w600)),
                    ),
                  ),
                  const SizedBox(height: AppSpacing.xl),
                ],
              ),
            ),
    );
  }
}

class _PreferenceSection extends StatelessWidget {
  const _PreferenceSection({
    required this.title,
    required this.description,
    required this.child,
  });

  final String title;
  final String description;
  final Widget child;

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: const EdgeInsets.all(AppSpacing.md),
      decoration: BoxDecoration(
        color: Theme.of(context).cardColor,
        borderRadius: BorderRadius.circular(AppRadius.large),
        boxShadow: [
          BoxShadow(
            color: Colors.black.withOpacity(0.04),
            blurRadius: 8,
            offset: const Offset(0, 2),
          ),
        ],
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(title,
              style: const TextStyle(
                  fontSize: 16, fontWeight: FontWeight.bold)),
          const SizedBox(height: 4),
          Text(description,
              style: TextStyle(
                  fontSize: 12, color: AppColors.textSecondary)),
          const SizedBox(height: AppSpacing.md),
          child,
        ],
      ),
    );
  }
}

class _ToggleChip extends StatelessWidget {
  const _ToggleChip({
    required this.label,
    required this.isSelected,
    required this.onTap,
  });

  final String label;
  final bool isSelected;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: onTap,
      child: Container(
        padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 8),
        decoration: BoxDecoration(
          color: isSelected
              ? AppColors.primary.withOpacity(0.1)
              : Theme.of(context).cardColor,
          borderRadius: BorderRadius.circular(AppRadius.medium),
          border: Border.all(
            color: isSelected
                ? AppColors.primary
                : AppColors.separator,
            width: isSelected ? 2 : 1,
          ),
        ),
        child: Text(
          label,
          style: TextStyle(
            fontSize: 14,
            color: isSelected
                ? AppColors.primary
                : AppColors.textPrimary,
            fontWeight:
                isSelected ? FontWeight.w600 : FontWeight.normal,
          ),
        ),
      ),
    );
  }
}
